### ISSUE TITLE
[css-multicol] What is the max-content width of a muticol container with only column-width:<length>

### ISSUE THREAD

**Issue by aethanyc** [w3c/csswg-drafts#9103]
This is from [bug 1844870](https://bugzilla.mozilla.org/show_bug.cgi?id=1844870). See [this testcase](https://bugzilla.mozilla.org/attachment.cgi?id=9345131).

Firefox 115 puts all the content in one column. Google Chrome 117 & Safari 17.0 create many short columns, and the multicol container's width is way larger than the browser window width.

The only relevant spec I found is a note in https://drafts.csswg.org/css-multicol/#pseudo-algorithm

> The used width U of the multi-column container can depend on the element’s contents, in which case it also depends on the computed values of the [column-count](https://drafts.csswg.org/css-multicol/#propdef-column-count) and [column-width](https://drafts.csswg.org/css-multicol/#propdef-column-width) properties. This specification does not define how U is calculated.

**Comment by dholbert** [comment]
The Blink/WebKit behavior seems to be a bug; they seem to be computing the max-content width of the element **as if it were not a multi-column container**.

Here's a testcase to demonstrate:
https://jsfiddle.net/dholbert/zne65cby/

markup in case the fiddle disappears:
```HTML
<!DOCTYPE html>
<style>
.container {
  border: 3px solid black;
  width: max-content;
  font: 10px monospace;
}

.multicol {
  column-width: 5em;
  margin-bottom: 5px;
  border-color: orange;
  column-rule: 1px solid red;
}
</style>
<div class="container">
  a b c
</div>
<div class="container multicol">
  a b c
</div>

<div class="container">
  a b c d e f g h i
</div>
<div class="container multicol">
  a b c d e f g h i
</div>

<div class="container">
  a b c d e f g h i j k l m n o p q r
</div>

<div class="container multicol">
  a b c d e f g h i j k l m n o p q r
</div>

<div class="container">
  a b c d e f g h i j k l m n o p q r s t u v w x y z
</div>
<div class="container multicol">
  a b c d e f g h i j k l m n o p q r s t u v w x y z
</div>
```

* Firefox consistently sizes the orange-bordered multicol elements to `5em` wide (the width of a single column).
* Chrome/Safari consistently size the orange-bordered multicol elements as being the same width as their black-bordered non-multicol sibling.  In other words, they're resolving `width:max-content` as if the element were not a multicol.

**Comment by dholbert** [comment]
Screenshot of my above-linked jsfiddle in Firefox:
![image](https://github.com/w3c/csswg-drafts/assets/426803/1f6d724a-fddb-44e4-8133-881578b0d472)
vs. Chrome:
![image](https://github.com/w3c/csswg-drafts/assets/426803/94e35fac-9fc8-4af5-9718-2e437e6ffc7c)
vs. Epiphany/gnome-web (using WebKit):
![image](https://github.com/w3c/csswg-drafts/assets/426803/3e50dc71-ce54-42b8-96cf-8cab3dea46f0)

**Comment by dholbert** [comment]
Here's another testcase to get at what browsers are doing here: https://jsfiddle.net/dholbert/5ak3xdnh/

This one has two multicol elements, which each have a single child.  In the first multicol, the child has the alphabet laid out on one line (characters separated by spaces). In the second, the characters are separated by `br` elements.

Chrome/WebKit render the first multicol element as having several columns (since they're using what-I-think-is-a-bogus measurement for the max-content width); and they render the second multicol as having one column (correctly). Firefox renders both multicols as having one column (correctly, I think).

Comparing the two multicol elements: the *child element* obviously has a larger max-content width in the first one, since it's got a long single line of content. But that doesn't mean the *multicol* element should also have a larger max-content width.  The multicol element's content-width should be essentially determined by the column-width and number-of-columns.  So if either of the multicols in this fiddle were going to be extra-wide, it should be the second one (where the child is taller and hence more capable of generating columns beyond the first), -- i.e. the opposite of the Chromium/WebKit behavior.  But really there's nothing forcing any fragmentation breaks here, so both of these multicol elements should have a single column, with a max-content width being the width of that column.

**Comment by rachelandrew** [comment]
Just looking at this, there [used to be a definition of the size of the container](https://www.w3.org/TR/2016/WD-css-sizing-3-20160510/#multicol-min-content) in Box Sizing (when it was still Intrinsic and Extrinsic Sizing) but it was removed and I don't think it has made it back into the level 4 spec. I think that's probably where this should be defined.

**Comment by mstensho** [comment]
This is a dup of https://github.com/w3c/csswg-drafts/issues/1742 isn't it?

**Comment by rachelandrew** [comment]
Been poking around some more at this, just to clarify:

With a:

- `column-width: <length>`
- `column-count: auto`
- `inline-size: max-content` on the multicol container

Firefox makes the max-content size the width of one column, and puts all content into that one column.

Chrome and WebKit make the max-content size the size it would be if we do inline-size: max-content in block layout. Then put the content into as many columns in as will fit.

I think Firefox is doing what the removed text says though not sure if that's what people would expect of a max-content multicol container.

Adding to the agenda to ask the following questions:

1. Should this be defined in Sizing or in Multicol (level 2 at this point)? The current multicol L1 spec states that another spec is expected to define it.
2. What behavior do we want when a multicol container has an inline-size of `max-content`?

Browsers do seem to be interoperable with `inline-size: max-content` and a `column-count: <integer>`. Example of both here: https://codepen.io/rachelandrew/pen/dyadbaL

**Comment by dholbert** [comment]
@rachelandrew yup, your description of the behaviors matches my observations in my previous comment (though you stated it more clearly than I -- thanks!)

RE what behavior we'd like here: the Firefox behavior is at least simple and understandable.   I'm happy to consider even-more-reasonable behaviors if they're not too complex/expensive, but I think the Chromium/WebKit behavior for the `column-count:auto` scenario ends up using a width that's pretty arbitrary and doesn't make much sense.  The arbitrariness is particularly noticeable when the content is all on one line, as demonstrated by my [above-linked jsfiddle](https://jsfiddle.net/dholbert/5ak3xdnh/), and in [this abbreviated version of Rachel's codepen](https://codepen.io/dholbert/pen/GRzxXKN). Screenshot of that pen in Chrome -- notice all the awkward blank space on the right, and the text wrapping at ~2.5 lines per column:
![image](https://github.com/w3c/csswg-drafts/assets/426803/0fb1facc-97e7-4f35-83cb-68052aaf2a77)
Chrome is sizing the container to the max-content width of the content-as-if-it-were-a-regular-block (i.e. the length of that line of text); and as you can see in the screenshot, that's not at all related to how wide the content actually ends up being, when laid out in a multicol setting with specified-width columns.  So it's not really a meaningful/useful max-content measurement for this container.

**Comment by alisonmaher** [comment]
I was thinking about this one some more, and I think the point from @bfgeek in the minutes is correct in that we would need to run layout to get the number of forced breaks needed.

One example where this would be necessary is with relation to flexbox fragmentation. According to the [spec](https://drafts.csswg.org/css-flexbox-1/#pagination), "In a row flex container, the [break-before](https://drafts.csswg.org/css-break-3/#propdef-break-before) and [break-after](https://drafts.csswg.org/css-break-3/#propdef-break-after) values on flex items are propagated to the flex line."

The main problem this poses is that there may be more than one forced break in the same line, so naively counting the number of forced breaks would be incorrect, since two forced breaks in the same line will only produce one true forced break before or after the current flex line.

Unless we run layout, we will not know which items are placed in which lines, and thus, where the forced breaks will apply to (and which will be ignored due to an earlier forced break in the same line). Flex lines are also not a part of the DOM, so we do some "magic" at layout time to apply these various breaks according to where the items ended up, meaning that knowing the exact number of columns created due to forced breaks in the case of flex fragmentation is impossible before layout.
